source/common/stats/thread_local_store.cc

```diff
@@ -204,29 +255,130 @@ Histogram& ThreadLocalStoreImpl::ScopeImpl::histogram(const std::string& name) {
   // See comments in counter(). There is no super clean way (via templates or otherwise) to
   // share this code so I'm leaving it largely duplicated for now.
   std::string final_name = prefix_ + name;
-  HistogramSharedPtr* tls_ref = nullptr;
+  std::unique_lock<std::mutex> lock(parent_.lock_);
+  ParentHistogramImplSharedPtr& central_ref = central_cache_.histograms_[final_name];
+
+  std::vector<Tag> tags;
+  std::string tag_extracted_name = parent_.getTagsForName(final_name, tags);
+  if (!central_ref) {
+    // Since MetricImpl only has move constructor, we are explicitly copying here.
+    std::string central_tag_extracted_name(tag_extracted_name);
+    std::vector<Tag> central_tags(tags);
+    central_ref.reset(new ParentHistogramImpl(final_name, parent_, *this,
+                                              std::move(central_tag_extracted_name),
+                                              std::move(central_tags)));
+  }
+  return *central_ref;
+}
+
+Histogram& ThreadLocalStoreImpl::ScopeImpl::tlsHistogram(const std::string& name) {
+  // See comments in counter() which explains the logic here.
+
+  // Here prefix will not be considered because, by the time ParentHistogram calls this method
```

Comment:
Thanks for pointing this out - could we document this in the base class header since it diverges from the behavior of the other 3 stat creation methods?